### PR TITLE
Email encoding fix with RFC 3986

### LINKED ISSUE #15480: Cannot log in using an email address containing a special character such as a single quote
Steps to reproduce:

1. Have an email address that has not been used before to create an account on WordPress.com and that contains a special character such as a single quote.
1. Open the app and log out if you were already logged in.
1. Tap the "Log in or sign up with WordPress.com" button.
1. Enter the email address from step 1.
1. Complete the account creation including any email verification required.
1. Open the app and log out again.
1. Try to log in using the email address from step 1.

**Results:** Signup allows me to use a single quote special character in an email address when I create an account but then I cannot log in when I try to use that email in the login flow. (<a href="https://cloudup.com/caQlMDo8fdx">6m57s</a>) 

**Note:** The error in this case is a bit odd since it mentions a problem with email that I don't think is relevant and shows the signup screen in the background even though an account already exists at that point for the email being tested.

**Error:** 

> There was some trouble sending the email. You can retry now or close and try again later. 

image

<sup>Tested signup using `design5279+android'test101821@gmail.com` using WPAndroid 18.3 Play Store version on Pixel 3 Android 11.</sup>

**App logs:** 

```
33 - [Oct-18 20:36 STATS] 🔵 Tracked: login_email_form_viewed
34 - [Oct-18 20:36 STATS] 🔵 Tracked: unified_login_step, Properties: {"source":"default","flow":"wordpress_com","step":"start"}
35 - [Oct-18 20:36 NUX] LoginEmailFragment: Google API client connected
36 - [Oct-18 20:36 STATS] 🔵 Tracked: unified_login_interaction, Properties: {"source":"default","flow":"wordpress_com","step":"start","click":"select_email_field"}
37 - [Oct-18 20:36 NUX] LoginEmailFragment: Autofill framework is enabled. Disabling hint picker dialog.
38 - [Oct-18 20:37 STATS] 🔵 Tracked: unified_login_interaction, Properties: {"source":"default","flow":"wordpress_com","step":"start","click":"submit"}
39 - [Oct-18 20:37 API] Dispatching action: AccountAction-FETCH_AUTH_OPTIONS
40 - [Oct-18 20:37 API] Volley error on https://public-api.wordpress.com/rest/v1.1/users/design5279+android'test101821@gmail.com/auth-options/?locale=en_GB - exception: null
41 - [Oct-18 20:37 API] StackTrace: com.android.volley.ClientError
	at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:199)
	at com.android.volley.NetworkDispatcher.processRequest(NetworkDispatcher.java:131)
	at com.android.volley.NetworkDispatcher.processRequest(NetworkDispatcher.java:111)
	at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:90)

42 - [Oct-18 20:37 API] Dispatching action: AccountAction-FETCHED_AUTH_OPTIONS
43 - [Oct-18 20:37 API] OnAuthOptionsFetched has error: UNKNOWN_USER - User does not exist.
44 - [Oct-18 20:37 API] Dispatching action: AuthenticationAction-SEND_AUTH_EMAIL
45 - [Oct-18 20:37 STATS] 🔵 Tracked: login_magic_link_open_email_client_viewed
46 - [Oct-18 20:37 STATS] 🔵 Tracked: unified_login_step, Properties: {"source":"default","flow":"signup","step":"magic_link_requested"}
47 - [Oct-18 20:37 API] Volley error on https://public-api.wordpress.com/rest/v1.1/auth/send-signup-email/?locale=en_GB - exception: null
48 - [Oct-18 20:37 API] StackTrace: com.android.volley.ClientError
	at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:199)
	at com.android.volley.NetworkDispatcher.processRequest(NetworkDispatcher.java:131)
	at com.android.volley.NetworkDispatcher.processRequest(NetworkDispatcher.java:111)
	at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:90)

49 - [Oct-18 20:37 API] Dispatching action: AuthenticationAction-SENT_AUTH_EMAIL
50 - [Oct-18 20:37 STATS] 🔵 Tracked: signup_magic_link_failed
51 - [Oct-18 20:37 API] OnAuthEmailSent error: USER_EXISTS - User with this email already has an account.
52 - [Oct-18 20:37 STATS] 🔵 Tracked: unified_login_failure, Properties: {"source":"default","flow":"signup","step":"magic_link_requested","failure":"There was some trouble sending the email. You can retry now or close and try again later."}
53 - [Oct-18 20:37 UTILS] App goes to background
54 - [Oct-18 20:37 STATS] 🔵 Tracked: application_closed, Properties: {"time_in_main_reader":2,"last_visible_screen":"Unknown","time_in_reader_paged_post":0,"time_in_app":113,"time_in_subfiltered_list":0,"time_in_reader_filtered_list":0}
```

**Workaround:** Log in to the affected WordPress.com account using a web browser, go to WP Admin > Profile > Get Apps and use the "Email me a log in link" to get a magic login link.

**Notes:** On the web, it's possible to log in by username instead of email address as an easy workaround. In the iOS app, login is completely blocked by this. In the Android app, there's actually a workaround which is to use the magic login link from the Get Apps page on the web so while it's still a problem for affected users it's not as urgent as it is for the iOS app.

<sup>(internal reference: p4a5px-2MC-p2 /hat tip @mdrockwell)</sup>

### PR DESCRIPTION
**Pull request by jd-alexander**
Fixes #15480

Related PRs:
- https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2171
- https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/73

## Description 

Encodes delimiters in email addresses when calling the `auth-options` endpoint, based on the [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-2.2) specification.

See: [Similar fix in iOS](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/457).

More info in this issue comment: https://github.com/wordpress-mobile/WordPress-Android/issues/15480#issuecomment-1202342661.

**To Test:**

**Setup** (feel free to skip if you already have a passwordless account)

0. For all steps, utilize an email with an apostrophe (such as `test'email@gmail.com`) and make sure you don't get an error when attempts are made to log in or sign up.
1. Log out if you are logged in.
1. On the Prologue screen, if the Smart Lock dialog appears, dismiss it.
1. Tap **Continue with WordPress.com**.
1. On the Get Started screen, enter an email address that is _not_ associated with a WordPress.com account.
1. Complete the sign-up flow normally.

**Test 1: Passwordless account / Magic Link login**

0. Clear app data.
1. On the Prologue screen, if the Smart Lock dialog appears, dismiss it.
1. Tap **Continue with WordPress.com**.
1. On the Get Started screen, enter an email address that is associated with a passwordless WordPress.com account.
1. Tap **Continue**.
1. Notice the Login Magic Link Sent screen.
1. Notice how there's no option to login with a password.
1. Tap **Check email**.
1. On the login email sent, tap **Log in to the app**.
1. Complete the login flow normally.

**Test 2: Account with password / Password login**

0. Clear app data.
1. On the Prologue screen, if the Smart Lock dialog appears, dismiss it.
1. Tap **Continue with WordPress.com**.
1. On the Get Started screen, enter an email address that is associated with a WordPress.com account that has a password set.
1. Tap **Continue**.
1. Notice the Password screen.
1. Notice how there's an option to login with a magic link.
1. Enter the WordPress.com account password.
1. Tap **Continue**.
1. Complete the login flow normally.

**Test 3: Sign up with an email not associated with a WordPress.com account**
0. Clear app data.
1. On the Prologue Screen, tap **Log In**.
2. If the Smart Lock dialog appears, dismiss it.
3. On the Email Screen, enter an email address that **_is not_** associated with a WordPress.com account.
4. Tap **Next**.
5. Notice the Sign-Up Magic Link screen.
6. Check email inbox of address used in Step 3.
7. Notice signup email from WordPress.com.
8. Complete the sign-up flow normally.

**Test 4: Check iOS & Android are in sync for emails having a `'` (optional)**
1. Create an account on an iOS with an e-mail having a single quote (E.g.: `main_email+test`ios@gmail.com`)
2. Login on iOS with that email
   a. Check under `Me` → `Account Settings` that the e-mail address is displayed correctly.
3. Login on Android with the same email
   a. **Expect** a successful login
   b. Check under `Me` → `Account Settings` that the e-mail address is displayed correctly.
4. Repeat steps 1-3 creating an account on `Android` first (E.g.: `main_email+test`android@gmail.com`)
   a. **Expect** same results.

## Previews

<details>
<summary>Before</summary>

https://user-images.githubusercontent.com/1509205/140247809-fa135c81-4557-427e-9308-2ec71f4198eb.mov
</details>

<details open>
<summary>After</summary>

https://user-images.githubusercontent.com/4588074/182418659-8275dae2-6dd8-4bb1-8352-e27c29debe18.mp4
</details>

## Merging Strategy

**Notice** ✅ One review should be enough to merge these PRs.

<details>
<summary>📂 Click to reveal merging workflow</summary>

1. Merge the  `WordPress-FluxC-Android` PR first
2. Wait for the [BuildKite Job for trunk](https://buildkite.com/automattic/wordpress-fluxc-android) to complete
   - Then go to the Job details in BuildKite and open subjob with name `Publish :fluxc`
   - Search for `is succesfully published` and **Copy** the value before that text, looking like  `trunk-{commit_sha}`.
   - **Paste** this value in the `./WordPressLoginFlow/build.gradle` file of `WordPress-Login-Flow-Android`, at:
```diff
@@ -55,7 +55,7 @@ dependencies {

api 'com.google.android.gms:play-services-auth:18.1.0'

-    implementation("org.wordpress:fluxc:2171-bac4c0dab01c00b98c2e88b7421ee5c84f408439") {
+    implementation("org.wordpress:fluxc:trunk-{commit_sha}") {
exclude group: "com.android.support"
exclude group: "org.wordpress", module: "utils"
}
```

3. Commit & push the change to the `WordPress-Login-Flow-Android` PR
4. Merge the `WordPress-Login-Flow-Android` PR
5. Also **Paste** the same FluxC version in the `./build.gradle` file of `WordPress-Android`, for `ext.fluxCVersion`.
6. Wait for the [BuildKite Job for trunk](https://buildkite.com/automattic/wordpress-login-flow-android) to complete
   - Then go to the Job details in BuildKite and open subjob with name `Publish to s3`
   - Search for `is succesfully published` and **Copy** the value before that text, looking like  `trunk-{commit_sha}`.
   - **Paste** this value in the `./build.gradle` file of `WordPress-Android`, for `ext.wordPressLoginVersion`.

7. Commit & push this change to the `WordPress-Android` PR
8. Merge the `WordPress-Android` PR.
</details>

## Regression Notes

1. Potential unintended areas of impact
Sign up and login-related API endpoints that provide authentication-based services.

6. What I did to test those areas of impact (or what existing automated tests I relied on)
I ran the login test cases that are shared above. The [SignUp](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/androidTest/java/org/wordpress/android/e2e/SignUpTests.java) and [LogIn](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/androidTest/java/org/wordpress/android/e2e/LoginTests.java) e2e tests were leveraged.

7. What automated tests I added (or what prevented me from doing so)
I didn't add any automated tests to cover this case since the fix is an optimization for the single quote edge case

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by AliSoftware** [comment]
(Moved to milestone 18.8 as I'm doing 18.7 code freeze today)

**Comment by AliSoftware** [comment]
@jd-alexander Any update on this one?

I'm submitting 18.9 today so this won't get in and be moved to next milestone once again, but this has been open for 2 months now so would be nice to get it moving (or closed) 🙂

**Comment by AliSoftware** [comment]
I'm about to do the 19.1 code freeze so I'll move this issue's milestone once again.

@jd-alexander (+ @develric as Groundskeeper this sprint): should we migrate this to `Future` milestone and/or add it to Groundskeeping queue, given that it has now been moved to next milestones multiple times without any new work on it?

**Comment by AliSoftware** [comment]
@jd-alexander @ravishanker This PR has been open and in Draft since Nov 2021, and I keep moving the milestone at each Code Freeze. It's probably worth either finally finishing it (using HACK week as an excuse to finally tackle it?) or set its milestone to Future / Groundskeeping projects?

**Comment by jd-alexander** [comment]
> @jd-alexander @ravishanker This PR has been open and in Draft since Nov 2021, and I keep moving the milestone at each Code Freeze. It's probably worth either finally finishing it (using HACK week as an excuse to finally tackle it?) or set its milestone to Future / Groundskeeping projects?

Hi @AliSoftware I was AFK for the last three months hence the late update. Thanks for all the bumps.  I set the milestone to Future. However, I am no longer working the WP codebase, I moved over to Tumblr so someone else can take this over and finalize it. 🙇🏾

**Comment by AliSoftware** [comment]
> Hi @AliSoftware I was AFK for the last three months hence the late update. Thanks for all the bumps. I set the milestone to Future. However, I am no longer working the WP codebase, I moved over to Tumblr so someone else can take this over and finalize it. 🙇🏾

Thanks for the update! I've pinged folks in our #wpmobile slack channel to check who could be interested in taking over this PR 🤞 . It seems quite close to being ready, so it would be a shame to let it fall into limbo — especially since it's addressing a Mobile Request by our users 🙃 

Enjoy your new role in Tumblr btw! 👋

**Comment by ovitrif** [comment]
As informed in https://github.com/wordpress-mobile/WordPress-Android/issues/15480#issuecomment-1202342661 the solution from this PR is not matching [the iOS fix](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/457) for issue https://github.com/wordpress-mobile/WordPress-Android/issues/15480.

On iOS we're encoding the single quote character to `%27`when calling the `auth-option` endpoint.

I'd like to work on adapting this fix to solve the issue also on Android in the same fashion it was solved for iOS, although I'm not entirely confident with the fix, given that on web it [isn't going to be possible](https://github.com/wordpress-mobile/WordPress-Android/issues/15480#issuecomment-1201345647) to login with an e-mail address containing a single quote.

**Comment by ovitrif** [comment]
Reopening after updating the FluxC version to the latest commit from https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2171.

---

ℹ️ The reason it closed automatically was a force push after rebasing on `trunk` and always approving `theirs` during merge conflict resolution.

**Review by mkevins — APPROVED** [review]
I've tested this and confirmed for my older account with a single quote in the address that it is working well without regression (the magic link still works). Also, the code looks good too. :+1:  Nice work Ovi!

**Comment by mkevins** [comment]
I have retested this with the updates to the Login library references, and it is working as expected. Let's merge it once the conflicts are resolved 👍
